### PR TITLE
npu gemm patch

### PR DESCRIPTION
**Pull request by a550580874**
[no_gemm_change.log](https://github.com/user-attachments/files/26931656/no_gemm_change.log)
[gemm_change.log](https://github.com/user-attachments/files/26931658/gemm_change.log)
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

This PR adds an NPU-only monkey patch for the MoE grouped GEMM path used by `fsdp2_moe.py`.

When running on Ascend/NPU, the patch replaces `transformers.integrations.moe._grouped_mm` with an NPU implementation backed by `torch_npu` grouped matmul. The goal is to improve MoE training performance on NPU while keeping numerical behavior aligned with the original implementation.

## What is changed

- add an NPU-only monkey patch for the MoE grouped GEMM path
- replace `transformers.integrations.moe._grouped_mm` with `_grouped_mm_npu` on NPU
- keep the default behavior unchanged on non-NPU environments
- add helper files for running `fsdp2_moe.py` with the NPU patch enabled

## Motivation

For MoE training on Ascend/NPU, the default grouped GEMM path is not optimal for NPU execution. This PR introduces a minimal NPU-specific patch so that the MoE grouped matmul path can use the native NPU grouped GEMM kernel, improving training performance while keeping numerical behavior aligned.

## Scope

This PR is intended for the NPU/Ascend path only.

- non-NPU behavior is unchanged
- the patch is applied only when NPU is detected
- the code change is localized and low-risk

### Accuracy alignment

The patched run is numerically aligned with the original run in the compared logs.

Checked steps show identical values for both:
- `loss`
- `grad_norm`

Examples:
- step 2: `loss=11.7626`, `grad_norm=2.818920`
- step 4: `loss=11.8967`, `grad_norm=3.034608`
- step 10: `loss=11.2776`, `grad_norm=4.786496`
- step 20: `loss=10.4782`, `grad_norm=4.824494`

This indicates that the patched NPU grouped GEMM path is numerically aligned with the original implementation for this experiment.

### Performance improvement

The patched run logs:
`[PATCH] transformers.integrations.moe._grouped_mm -> _grouped_mm_npu`

After excluding the first warmup step:
- baseline (without patch): average step time is about **1.74 s/step**
- patched version: average step time is about **0.46 s/step**

This gives an approximate **3.77x speedup** on the tested run.

A few concrete step time examples:

Baseline:
- `1.5999`
- `2.2731`
- `1.6419`
- `1.6620`

Patched:
- `0.4250`
- `0.9911`
- `0.4147`
- `0.4503`

The first step is slower in the patched run due to initialization and warmup overhead, but the steady-state step time is significantly lower.

## Validation

Validation for this PR is based on:
1. successful patch activation on NPU
2. identical `loss` and `grad_norm` on matched steps in the compared logs
3. significantly reduced steady-state `step_time`

## Notes

This PR focuses on the NPU MoE grouped GEMM path only. It does not change the default behavior for non-NPU backends.

### CODE REVIEW

File: src/twinkle/kernel/monkey_patch_npu.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import functools
```
Comment (tardis-key) [inline src/twinkle/kernel/monkey_patch_npu.py:1]: Refer to VERL, aggregate all patches in this file, and perform full replacement externally only after is_npu_available() at twinkle initialization.

File: cookbook/transformers/fsdp2_moe_npu.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+# CANN loading
+source /usr/local/Ascend/ascend-toolkit/set_env.sh
+
+# export CPU_AFFINITY_CONF=2
```
Comment (tardis-key) [inline cookbook/transformers/fsdp2_moe_npu.sh]: What are these variables used for? Please confirm if they need to be kept.

File: cookbook/transformers/fsdp2_moe.py [hunk 3]
```diff
@@ -17,6 +18,13 @@
 logger = get_logger()
 
 
+# npu patch
+if is_torch_npu_available():
+    from monkey_patch_npu import apply_hf_moe_grouped_mm_patch
+    apply_hf_moe_grouped_mm_patch()
+    import torch_npu
+    from torch_npu.contrib import transfer_to_npu
```
Comment (tardis-key) [inline cookbook/transformers/fsdp2_moe.py]: maybe we can move transfer_to_npu into monkey_patch_npu.py if it's necessary for patching

**Comment by tardis-key** [comment]
lgtm

File: src/twinkle/kernel/monkey_patch_npu.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        assert group_list.dim() == 1, f"group_list must be [E], got {tuple(group_list.shape)}"
+        assert weight_ekn.dim() == 3, f"weight_ekn must be [E, K, N], got {tuple(weight_ekn.shape)}"
+        assert group_list.numel() == weight_ekn.size(0), (
+            f"group_list len {group_list.numel()} != num_experts {weight_ekn.size(0)}"
+        )
+        assert x.size(1) == weight_ekn.size(1), (
+            f"input dim mismatch: x.shape={tuple(x.shape)}, weight_ekn.shape={tuple(weight_ekn.shape)}"
+        )
+
+        group_list = group_list.to(torch.int64)
```
[the remark this answers is not in the document]
Comment (a550580874) [reply src/twinkle/kernel/monkey_patch_npu.py:18]: Since we can't guarantee all potential upstream callers of group_list have already performed the int64 conversion, I kept this explicit cast to ensure type safety defensively, regardless of where it's passed from.

File: src/twinkle/kernel/monkey_patch_npu.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    assert weight_ekn.dim() == 3, f"weight_ekn must be [E, K, N], got {tuple(weight_ekn.shape)}"
+    assert offs.dim() == 1, f"offs must be [E], got {tuple(offs.shape)}"
+    assert weight_ekn.size(0) == offs.numel(), (
+        f"weight_ekn.size(0)={weight_ekn.size(0)} != offs.numel()={offs.numel()}"
+    )
+
+    counts = torch.empty_like(offs)
+    counts[0] = offs[0]
+    if offs.numel() > 1:
+        counts[1:] = offs[1:] - offs[:-1]
```
[the remark this answers is not in the document]
Comment (a550580874) [reply src/twinkle/kernel/monkey_patch_npu.py:69]: In-place operations on tensor views (like `counts[1:] -= ...`) can sometimes trigger sub-optimal memory patterns, implicit contiguous copies, or kernel dispatch fallbacks. Explicitly allocating memory and avoiding in-place view modifications is safer and more predictable for the hardware.

File: src/twinkle/kernel/monkey_patch_npu.py (hunk 1 above)
Comment (a550580874) [reply src/twinkle/kernel/monkey_patch_npu.py:1]: Moved all patches into apply_npu_patch within monkey_patch_npu.

File: cookbook/transformers/fsdp2_moe_npu.sh (hunk 2 above)
Comment (a550580874) [reply cookbook/transformers/fsdp2_moe_npu.sh]: Removed

File: src/twinkle/utils/utils.py [hunk 6]
```diff
@@ -77,3 +77,12 @@ def should_exclude_file(file_path, file_name):
                     destination = os.path.join(dest_dir, file_name)
                     if not os.path.exists(destination):
                         shutil.copy2(file_path, destination)
+
+
+def is_torch_npu_available():
```
Comment (tastelikefeet) [inline src/twinkle/utils/utils.py]: 这里不需要重新写，使用twinkle.utils.Torch.is_npu_available()

File: cookbook/transformers/fsdp2_moe.py (hunk 3 above)
Comment (a550580874) [reply cookbook/transformers/fsdp2_moe.py]: removed

File: src/twinkle/utils/utils.py (hunk 6 above)
Comment (a550580874) [reply src/twinkle/utils/utils.py]: repaired

**Review by tastelikefeet — APPROVED** [review]
